File: ArraysQuestions/IN_CPP/Container_with_most_water.cpp
/*
    Container with Most Water:

    Examples :
        Input: array = [1, 5, 4, 3]
        Output: 6
        Explanation :
            5 and 3 are distance 2 apart.
            So the size of the base = 2.
            Height of container = min(5, 3) = 3.
            So total area = 3 * 2 = 6

        Input: array = [3, 1, 2, 4, 5]
        Output: 12
        Explanation :
            5 and 3 are distance 4 apart.
            So the size of the base = 4.
            Height of container = min(5, 3) = 3.
            So total area = 4 * 3 = 12
*/

#include <iostream>
#include <vector>
using namespace std;

// Time complexity O(n^2)
int maxArea1(vector<int> &height)
{
    int area = 0;
    for (int i = 0; i < height.size(); i++)
    {
        for (int j = i + 1; j < height.size(); j++)
        {
            area = max(area, min(height[j], height[i]) * (j - i));
        }
    }

    return area;
}

// Time Complexity O(n)
int maxArea2(vector<int> height)
{
    int left = 0, right = height.size() - 1;
    int area = 0;

    while (left < right)
    {
        area = max(area, min(height[left], height[right]) * (right - left));

        if (height[left] < height[right])
        {
            left++;
        }
        else
        {
            right--;
        }
    }
    return area;
}

int main()
{
    vector<int> height;
    height.push_back(1);
    height.push_back(8);
    height.push_back(6);
    height.push_back(2);
    height.push_back(5);
    height.push_back(4);
    height.push_back(8);
    height.push_back(3);
    height.push_back(7);

    cout << maxArea1(height) << endl;
    cout << maxArea2(height) << endl;
}

File: Questions_In_Java/ArraysQuestions/ContainerWithMostWater.java
package ArraysQuestions;

/*
    Container with Most Water:

    Examples :
        Input: array = [1, 5, 4, 3]
        Output: 6
        Explanation :
            5 and 3 are distance 2 apart.
            So the size of the base = 2.
            Height of container = min(5, 3) = 3.
            So total area = 3 * 2 = 6

        Input: array = [3, 1, 2, 4, 5]
        Output: 12
        Explanation :
            5 and 3 are distance 4 apart.
            So the size of the base = 4.
            Height of container = min(5, 3) = 3.
            So total area = 4 * 3 = 12
*/

import java.util.ArrayList;

public class ContainerWithMostWater {
    public static void main(String[] args) {
        ArrayList<Integer> height = new ArrayList<>();
        height.add(1);
        height.add(8);
        height.add(6);
        height.add(2);
        height.add(5);
        height.add(4);
        height.add(8);
        height.add(3);
        height.add(7);

        System.out.println("Max Area is : " + maxArea1(height));
    }

    // Time complexity O(n^2)
    private static int maxArea1(ArrayList<Integer> arr) {
        int area = 0;
        for (int i = 0; i < arr.size(); i++) {
            for (int j = i+1; j < arr.size(); j++) {
                area = Math.max(area, Math.min(arr.get(j), arr.get(i)) * (j - i));
            }
        }
        return area;
    }
}
